### PR TITLE
feat(@astrojs/cloudflare): Add support for wasm module imports

### LINKED ISSUE #8541: Cloudflare adapter doesn't support WASM, despite natively supporting  WASM modules
### Astro Info

```block
Astro                    v3.0.12
Node                     v20.6.1
System                   macOS (arm64)
Package Manager          npm
Output                   server
Adapter                  @astrojs/cloudflare
Integrations             none
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When attempting to import and bundle a wasm module, the cloudflare adapter throws errors from vite and esbuild, despite wasm module import [being possible in native cloudflare worker imports](https://developers.cloudflare.com/workers/runtime-apis/webassembly/javascript/#use-from-javascript).

When attempting to import a wasm file, you'll see the following error during an `astro dev` server:

```
Failed to parse source for import analysis because the content contains invalid JS syntax. You may need to install appropriate plugins to handle the .wasm file format, or if it's an asset, add "**/*.wasm" to `assetsInclude` in your configuration.
```

This appears to be coming from the initial vite build. Adding vite plugin for wasm, [such as this](https://github.com/Aslemammad/vite-plugin-cloudflare), and marking the import as a `.wasm?module` is promising: it gets the astro dev server running. However when running astro build, further errors are thrown when esbuild doesn't know what to do with the `.wasm` import when bundling.

```
$ npx astro build
# ...
✘ [ERROR] Could not resolve "./../../_astro/add.f61fd62f.wasm"

    dist/$server_build/chunks/pages/index_0ed3720d.mjs:1:17:
      1 │ import init from './../../_astro/add.f61fd62f.wasm';
```

### What's the expected result?

It seems like wasm module style imports should be relatively straight forward to support since they are native to cloudflare. From what I've tried, its also the only way in cloudflare to import a wasm file. (loading a fetched file from binary throws errors in the cloudflare runtime). 

Additionally, adding support for wasm in cloudflare should also help add sharp support to the cloudflare adapter in the future [once there's a web assembly build of sharp](https://github.com/lovell/sharp/pull/3522)

### Link to Minimal Reproducible Example

https://github.com/adrianlyjak/astro/tree/1d780c08b44184a2306a8ba181160e4928a88cd6/packages/integrations/cloudflare/test/fixtures/wasm

### Participation

- [X] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by adrianlyjak**
## Changes

- Adds vite and esbuild plugins for loading web assembly modules to fix #8541

## Testing

- automated tests were added that imports web assembly modules from the 3 different possible directory structures
- manual testing can be done be importing a simple .wasm file and using it. Wasm files should be able to be imported from node_modules as well. 
- Refactors cloudflare wrangler tests to be more reliable by dynamically rotating ports, and retrying timeouts. Removes `kill-port` dependency. I attempted to removes skips for flaky startup (I wasn't encountering any flakiness after the port fix locally), however the windows and node 20 environments seem to still consistently fail on 1 or 2 wrangler cli tests. Instead I refactored the skip behavior to be a little simpler to opt-in or out.

## Docs

Updated README for cloudflare adapter with small section on wasm modules.

### CODE REVIEW

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [reply packages/integrations/cloudflare/src/index.ts]: @alexanderniebuhr could you explain why, please? I'm curious :)

Comment (adrianlyjak) [reply packages/integrations/cloudflare/src/index.ts]: I'll add my 2 cents. It would be nice if this was zero configuration. Its not really "changing" behavior like other configs, rather its adding behavior that the user would expect to work given what works in cloudflare.

That being said, this multiple esbundle depending on depth of the entrypoint is a bit of a hack, and if it stays that way, adding a configuration to opt-in to wasm loading for at least the function per route case makes sense to me.

@alexanderniebuhr were your reservations around this hack or something else?

About the hack, In my development, I wasn't able to specify a different import pattern for the different entrypoints using its `onResolve` callback. But it seems like there must be some other better way to do it. I was just hoping that someone might have better insight. Looking deeper at the esbuild plugin documentation, I might be able to do something with the [`onEnd` callback](https://esbuild.github.io/plugins/#on-end) to modify the imports within each entry

Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts]: @ematipico @adrianlyjak let me explain!

I'm genuinely excited about this feature and would love to see it enabled by default in the long run. 🚀

However, I've got a few thoughts:

- I'm a bit hesitant about the `entryPathsGroupedByDepth` solution. I believe there might be a more efficient approach out there. I'm already thinking about it but need some more time, to fully understand the requirements.
- However I don't think it should block this PR - we should iterate on it in a future PR!
- Given that `wasm` isn't a typical use-case for majority of users, my suggestion would be to make this feature opt-in for now. This way, we ensure we don't inadvertently degrade the experience for users without `wasm` - which we test for, but it still feels a bit hacky.
- I also haven't dived deeper into a thorough code review yet. So I don't really know if the current solution is that hacky or fine. I was just going by my initial gut feeling and the comment provided by @adrianlyjak.
- Once we've tested this with real-world use cases & optimized the adapter to reduce code branches _(I'm working on that)_, I would consider making the default `true`, but still leave the flag intact _(for user choice)_.

This mirrors how I usually approach runtime support: opt-in first & switching to default once I'm sure it works for 95% of users.

Comment (adrianlyjak) [reply packages/integrations/cloudflare/src/index.ts]: Yeah, I'm looking a little deeper at the hack, and I think I may have just overlooked some things that should be able to remove the `entryPathsGroupedByDepth`. I'll let you know once that's updated.

I've also implemented a flag. That makes sense to slow roll this in case there is potential any potential fallout from the new functionality.

Comment (adrianlyjak) [reply packages/integrations/cloudflare/src/index.ts]: > However I don't think it should block this PR - we should iterate on it in a future PR!

Well, the [onEnd](https://esbuild.github.io/plugins/#on-end) callback looks promising, but still doesn't appear trivial to implement, so perhaps yes, given the opt-in, deferring that optimization seems like a better decision.

I'm currently running into another head scratcher. This doesn't work well in hybrid mode where a client side page imports wasm. I've added a failing test.

Currently, due to the vite build, it's just rewriting to a `*.wasm` import for cloudflare workers. However, when generating the static pages, that doesn't work, giving the following opaque error.

```
[vite:wasm-module-loader] Could not load /Users/adrianlyjak/dev/futz/astro/packages/integrations/cloudflare/test/fixtures/wasm/src/util/add.wasm?module (imported by test/fixtures/wasm/src/pages/hybrid.ts): WASM module's cannot be loaded unless you add `wasmModuleImports: true` to your astro config.
```

I tried tweaking the vite/rollup plugin. The SSG build isn't in a node-like environment, but I can embed the wasm as a base64 string. The unfortunate thing here though is it seems like the SSG mode and SSR mode use the same vite build, so the vite build is somewhat at a stand still. Cloudflare only supports importing .wasm modules, and the ssg builder does not. And unfortunately this only shows up at build time, so the dev server is misleading that things will work.

The only further idea to support both would be to use base64 in the vite build always, and then somehow re-interpret an wasm import within the worker esbuild.

Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts]: @adrianlyjak I'm keen to assist and offer a fresh perspective. While I haven't fully grasped all the issues yet, I'll delve deeper into your implementation to better understand the challenges you've described. I believe there might be some hooks we can leverage to streamline the process. To reiterate my earlier point, I feel we should merge this PR as it stands and focus on refining it in subsequent iterations.

What are your thoughts, @ematipico?

Comment (adrianlyjak) [reply packages/integrations/cloudflare/src/index.ts]: > The only further idea to support both would be to use base64 in the vite build always, and then somehow re-interpret an wasm import within the worker esbuild.

I ended up doing this. It seems to work well and not be too complicated. Main complaint is that there's loosely coupled and hard to follow assumptions between the 2 build stages

Comment (ematipico) [reply packages/integrations/cloudflare/src/index.ts]: Given the long discussion and the technical details that surfaced, I think we should keep the feature opt-in, so we can discuss them offline (discord, github issue, etc.) I want @adrianlyjak 's contribution to be merged ASAP, but we shouldn't rush into "enabled by default".

Comment (adrianlyjak) [reply packages/integrations/cloudflare/src/index.ts]: Currently the feature is opt-in. What's the process for merging? Or do you want further code review? I see this is starting to accumulate conflicts, and I'm hoping to not have to keep up with those. I'll rebase this for the time being (excited about the node compatibility that's conflicting here 😄 )

Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts]: For me it looks good!
In general the process is other reviews, including docs. Once they are approve, merging will be handled by one of the maintainers :)

**Review by matthewp — APPROVED** [review]

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:198]:
```suggestion
default: `false`

Whether or not to import `.wasm` files [directly as ES modules](https://github.com/WebAssembly/esm-integration/tree/main/proposals/esm-integration). 

Add `wasmModuleImports: true` to `astro.config.mjs` to enable in both the Cloudflare build and the Astro dev server. 
```
This seems more in keeping with the option directly above, stating the default and firstly, defining the configuration setting itself.

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:201]:
```suggestion
// astro.config.mjs
import {defineConfig} from "astro/config";
```
We do add a file name whenever it's easy to do so, so that it's clear to someone just skimming code samples where this code is!

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:212]:
```suggestion
Once enabled, you can import a web assembly module in Astro with a `.wasm?module` import. 

The following example shows... (importing ... then ...)
```
Just a reminder that proper names, when we just mention them in a sentence, should be capitalized.

Ideally, we prefer to introduce code samples with a sentence like, "The following example shows... " and then an explanation of exactly what the code is doing. This means people don't have to both figure out what this example is, and then also how the thing you're demonstrating also works. We help them by telling them before they see the example that this is what the code accomplishes.

Also, the fact that it works in both modes is great to include! But, it seems out of place here. I suggest moving this sentence below the code sample, since it's not actually related to this code example.

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:228]:
```suggestion
The Cloudflare integration supports WASM module imports in both `server` and `hybrid` mode for a consistent development experience.
```
This seems more appropriate as a "final note" rather than interrupting explaining usage.

**Review by sarah11918 — COMMENTED** [review]
Thanks so much for these docs contributions, @adrianlyjak! I found them super easy to follow and I think they'll be helpful to people using the integration!  🙌 

I just had some suggestions for your consideration to adjust them a tiny bit so they would be more in line with how we'd normally write our Astro documentation. See what you think and as I always have to say, I'm just editing for style and structure. It's your job to make the statements true!  😉

File: packages/integrations/cloudflare/README.md
Comment (adrianlyjak) [reply packages/integrations/cloudflare/README.md]: Good points. I made the capitalization of `Wasm` consistent with that suggested at [https://webassembly.org/](https://webassembly.org/). And updated example introduction text to be more descriptive

File: packages/integrations/cloudflare/README.md
Comment (adrianlyjak) [reply packages/integrations/cloudflare/README.md]: Reading this sentence over, it felt redundant. I would imagine that user expectation is that the feature works out of the box. 

Updated instead to give an idea of a more exciting potential use case for Wasm. I like the addition example because its simple and intuitive, but its not a practical use Wasm.

**Review by ematipico — APPROVED** [review]
Great work!

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:219]:
```suggestion
The following is an example of importing a Wasm module that then responds to requests by adding the request's number parameters together. 
```
Nicely done, but even just this is sufficient! It's just about providing a little context!

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:235]:
```suggestion
While this example is trivial, Wasm can be used to accelerate computationally intensive operations which do not involve significant I/O such as embedding an image processing library.
```
That just wasn't a full sentence on its own so reworded slightly!

**Review by sarah11918 — COMMENTED** [review]
Looking good @adrianlyjak ! Just two tiny suggestions, then we're good to go!

File: packages/integrations/cloudflare/README.md
Comment (adrianlyjak) [reply packages/integrations/cloudflare/README.md]: @sarah11918 thanks! Updated

**Review by sarah11918 — APPROVED** [review]
Love to approve great docs! :partying_face:

**Pull request merged**
